### PR TITLE
add back method_missing support to set_unless

### LINKED ISSUE #5102: set_unless in dot notation causes Chef crashes
## Description

When attributes are used in dotted notation Chef produces crash with `default_unless`:
```
default_unless.foo = "bar"
```

## Chef Version

I see this crash on 12.12.13 and 12.12.15 but don't see it on 12.11.18

## Platform Version

Ubuntu 14.04

## Replication Case

```
root@vm108842:~# chef-client
Starting Chef Client, version 12.12.15
resolving cookbooks for run list: ["foo"]
Synchronizing Cookbooks:
  - foo (0.2.14)
Installing Cookbook Gems:
Compiling Cookbooks...

================================================================================
Recipe Compile Error in /var/chef/cache/cookbooks/foo/attributes/default.rb
================================================================================

NoMethodError
-------------
undefined method `foo=' for #<Chef::Decorator::Unchain:0x00000006cc1860>

Cookbook Trace:
---------------
  /var/chef/cache/cookbooks/foo/attributes/default.rb:1:in `from_file'

Relevant File Content:
----------------------
/var/chef/cache/cookbooks/foo/attributes/default.rb:

  1>> default_unless.foo = 'bar'
  2:

Platform:
---------
x86_64-linux


Running handlers:
  - MailHandler
  - MailHandler
Running handlers complete
Chef Client failed. 0 resources updated in 20 seconds
```

To reproduce simply create cookbook `foo` with the above line in `attributes/default.rb`

## Client Output

```
root@vm108842:~# chef-client -l debug
Starting Chef Client, version 12.12.15
resolving cookbooks for run list: ["foo"]
Synchronizing Cookbooks:
  - foo (0.2.14)
Installing Cookbook Gems:
Compiling Cookbooks...

================================================================================
Recipe Compile Error in /var/chef/cache/cookbooks/foo/attributes/default.rb
================================================================================

NoMethodError
-------------
undefined method `foo=' for #<Chef::Decorator::Unchain:0x0000000360c6d0>

Cookbook Trace:
---------------
  /var/chef/cache/cookbooks/foo/attributes/default.rb:1:in `from_file'

Relevant File Content:
----------------------
/var/chef/cache/cookbooks/foo/attributes/default.rb:

  1>> default_unless.foo = 'bar'
  2:

Platform:
---------
x86_64-linux


Running handlers:
  - MailHandler
  - MailHandler
Running handlers complete
Chef Client failed. 0 resources updated in 04 seconds
````

### PR DESCRIPTION
**Pull request by lamont-granquist**
closes #5102

### CODE REVIEW

**Comment by lamont-granquist** [comment]
@chef/client-core regression, needs reviewage

**Comment by btm** [comment]
👍

File: lib/chef/decorator/unchain.rb [hunk 1]
```diff
@@ -38,6 +38,22 @@ def []=(key, value)
         __path__.push(key)
         @delegate_sd_obj.public_send(__method__, *__path__, value)
       end
+
+      # unfortunately we have to support method_missing for node.set_unless.foo.bar = 'baz' notation
+      def method_missing(symbol, *args)
+        if symbol == :to_ary
+          merged_attributes.send(symbol, *args)
+        elsif args.empty?
+          Chef.log_deprecation %q{"method access to node attributes (node.foo.bar) is deprecated and will be removed in Chef 13, please use bracket syntax (node["foo"]["bar"])}
```
Comment (coderanger) [inline lib/chef/decorator/unchain.rb]: Rouge `"` at the start of the message from copy pasta?

**Comment by coderanger** [comment]
:+1: Other than message fix :)

File: lib/chef/decorator/unchain.rb (hunk 1 above)
Comment (lamont-granquist) [inline lib/chef/decorator/unchain.rb]: Heh, copied from master:  https://github.com/chef/chef/blob/master/lib/chef/node/attribute.rb#L520

**Pull request merged**